File: week-01/dahyen0o/12015.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static StringBuilder sb;

    static int N;
    static int[] array, lis;

    public static void main(String[] args) throws Exception {
      
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        sb = new StringBuilder();

        N = Integer.parseInt(br.readLine());
        
        array = new int[N];
        st = new StringTokenizer(br.readLine());

        for (int n = 0; n < N; n++) {
            array[n] = Integer.parseInt(st.nextToken());
        }

        lis = new int[N];
        Arrays.fill(lis, Integer.MAX_VALUE);

        int lisIdx = 0;

        lis[lisIdx++] = array[0];

        for (int n = 1; n < N; n++) {
            if (array[n] > lis[lisIdx - 1]) {
                lis[lisIdx++] = array[n];
            } else {
                lis[binarySearch(0, lisIdx, array[n])] = array[n];
            }
        }

        bw.write(sb.append(lisIdx).toString());
        bw.flush();
    }

    private static int binarySearch(int left, int right, int key) {
        int mid = 0;
        while (left < right) {
            mid = (left + right) / 2;
            if (lis[mid] < key) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return right;
    }
}

File: week-01/pushedrumex/12015.py
N = int(input())
arr = list(map(int, input().split()))

LIS = [arr[0]]

for n in arr:
    num = LIS[-1]
    if n > num: LIS.append(n)
    else:
        # n 을 포함했을 경우, 본인의 자리를 찾아 교환
        left, right = 0, len(LIS) - 1
        while left <= right:
            mid = (left + right) // 2
            if LIS[mid] < n:
                left = mid + 1
            else:
                result = mid
                right = mid - 1
        LIS[mid] = n

print(len(LIS))
